lib/src/main/java/org/esbtools/eventhandler/PromiseOfPromise.java

```diff
@@ -0,0 +1,69 @@
+/*
+ *  Copyright 2016 esbtools Contributors and/or its affiliates.
+ *
+ *  This file is part of esbtools.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.esbtools.eventhandler;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class PromiseOfPromise<U> implements Promise<U> {
```

Comment:
This is a refactoring which:

- Makes having a `thenPromise` (or equivalent) unnecessary
- Makes it more difficult (if not impossible) to lazily execute computation while in the Future domain
- Is less obvious how to use: I already have a promise from `request`... why do I need to forward to another?

I still don't see a compelling reason why an additional `thenPromise` API is so egregious. It is a situation--returning another Future from a callback on a prior Future--which is probably more canonical than not, and you could argue that the default `then` API should be one that works this way, and we instead use another API for callbacks which do _not_ return a Future themselves (maybe like, `finishWith(Callback)` or something of the sort). For instance, Guava has this same feature with `AsyncFunction`. Perhaps we could refactor to be more like the Guava API.